src/main/java/io/confluent/kafkarest/SimpleConsumerPool.java

```diff
@@ -34,37 +35,192 @@
 
-  public SimpleConsumerPool(int maxPoolSize, int poolInstanceAvailabilityTimeoutMs,
-                            Time time, SimpleConsumerFactory simpleConsumerFactory) {
+  private final ConsumerFactory<byte[], byte[]> simpleConsumerFactory;
+  private final Map<TopicPartition, Map<Long, List<Consumer<byte[], byte[]>>>> topicPartitionOffsetConsumers;
+  private int consumerSize;
+
+  public SimpleConsumerPool(int maxPoolSize,
+                            int poolInstanceAvailabilityTimeoutMs,
+                            int maxPollTime,
+                            Time time, ConsumerFactory<byte[], byte[]> simpleConsumerFactory) {
     this.maxPoolSize = maxPoolSize;
     this.poolInstanceAvailabilityTimeoutMs = poolInstanceAvailabilityTimeoutMs;
     this.time = time;
+    this.maxPollTime = maxPollTime;
+
     this.simpleConsumerFactory = simpleConsumerFactory;
 
-    simpleConsumers = new HashMap<String, SimpleConsumer>();
-    availableConsumers = new LinkedList<String>();
+    // using LinkedHashMap guarantees FIFO iteration order
+    topicPartitionOffsetConsumers = new LinkedHashMap<>();
+    consumerSize = 0;
+  }
+
+  public RecordsFetcher getRecordsFetcher(TopicPartition topicPartition) {
+    return new RecordsFetcher(topicPartition);
+  }
+
+  public final class RecordsFetcher implements AutoCloseable {
+
+    private Consumer<byte[], byte[]> consumer;
+    private final SimpleConsumerPool ownerPool;
+    private final long maxPollTime;
+    private boolean initialized;
+    private Time time;
+    private final TopicPartition topicPartition;
+
+    RecordsFetcher(TopicPartition topicPartition) {
+      this.ownerPool = SimpleConsumerPool.this;
+      this.time = SimpleConsumerPool.this.time;
+      this.maxPollTime = SimpleConsumerPool.this.maxPollTime;
+      this.topicPartition = topicPartition;
+      initialized = false;
+    }
+
+    // support for lazy initialization gives opportunity
+    // to add some kind of cache in future.
+    private Consumer<byte[], byte[]> initializeConsumer(long offset) {
+      if (!initialized) {
+        consumer = ownerPool.get(topicPartition, offset);
+        initialized = true;
+      }
+      return consumer;
+    }
+
+    public List<ConsumerRecord<byte[], byte[]>> poll(final long offset, final long count) {
+      if (count == 0) {
+        return Collections.emptyList();
+      }
+
+      List<ConsumerRecord<byte[], byte[]>> result = new ArrayList<>();
+
+      long endTime = time.milliseconds() + maxPollTime;
+      long pollTime;
+
+      Iterator<ConsumerRecord<byte[], byte[]>> it = null;
+      boolean enough = false;
+
+      while (!enough &&
+          (pollTime = endTime - time.milliseconds()) > 0) {
+
+        if (it == null || !it.hasNext()) {
+
+          if (!initialized) {
+            initializeConsumer(offset);
+          }
+
+          ConsumerRecords<byte[], byte[]> records = consumer
+              .poll(Math.max(0, pollTime));
+          if (records.isEmpty()) {
```

Comment:
This isn't necessary as the behavior is the same even if you get the iterator and execute the loop below (which will just exit immediately).